openedx_learning/core/components/models.py

```diff
@@ -356,48 +393,97 @@ class Meta:
+    app-extended models get deleted (e.g. if they are deprecated and removed).
+    """
+    # 100K is our limit for text data, like OLX. This means 100K *characters*,
+    # not bytes. Since UTF-8 encodes characters using as many as 4 bytes, this
+    # couled be as much as 400K of data if we had nothing but emojis.
+    MAX_TEXT_LENGTH = 100_000
+
+    raw_content = models.OneToOneField(
+        RawContent,
+        on_delete=models.CASCADE,
+        primary_key=True,
+        related_name='text_content',
+    )
+    text = models.TextField(null=False, blank=True, max_length=MAX_TEXT_LENGTH)
+    length = models.PositiveIntegerField(null=False)
 
-class ComponentVersionContent(models.Model):
+
+class ComponentVersionRawContent(models.Model):
     """
-    Determines the Content for a given ComponentVersion.
+    Determines the RawContent for a given ComponentVersion.
 
     An ComponentVersion may be associated with multiple pieces of binary data.
     For instance, a Video ComponentVersion might be associated with multiple
     transcripts in different languages.
 
-    When Content is associated with an ComponentVersion, it has some local
+    When RawContent is associated with an ComponentVersion, it has some local
     identifier that is unique within the the context of that ComponentVersion.
     This allows the ComponentVersion to do things like store an image file and
     reference it by a "path" identifier.
 
-    Content is immutable and sharable across multiple ComponentVersions and even
-    across LearningPackages.
+    RawContent is immutable and sharable across multiple ComponentVersions and
+    even across LearningPackages.
     """
-
+    raw_content = models.ForeignKey(RawContent, on_delete=models.RESTRICT)
     component_version = models.ForeignKey(ComponentVersion, on_delete=models.CASCADE)
-    content = models.ForeignKey(Content, on_delete=models.RESTRICT)
+
+    uuid = immutable_uuid_field()
     identifier = identifier_field()
 
+    # Is this RawContent downloadable during the learning experience? This isn't
+    # about public vs. private permissions on course assets, as that will be a
+    # policy that can be changed independently of new versions of the content.
+    # For instance, a course team could decide to flip their course assets from
+    # private to public for CDN caching reasons, and that should not require
+    # new ComponentVersions to be created.
+    # 
+    # What the ``downloadable`` field refers to is whether this asset is
+    # supposed to *ever* be directly downloadable by browsers during the
+    # learning experience. For instance, Problem Block OLX will contain the
+    # answers to the problem. The XBlock runtime and ProblemBlock will use this
+    # information to generate HTML and grade responses, but the the user's
+    # browser is never permitted to actually download the raw OLX itself.
+    downloadable = models.BooleanField(default=False)
```

Comment:
@bradenmacdonald, @kdmccormick, @feanil: This is where I think I landed in terms of the download-permission awareness of this layer of the code (which is to say, not much).

Reply:
Would `learner_downloadable` be more a accurate name? Because I'm assuming that it's still always downloadable by course authors for example. Based on the first line of the comment, that is what you have in mind.

Reply:
Also, I don't really understand; with this approach, if there is some PDF asset that the LMS wants to make available only to enrolled learners, how would we do that? Set `downloadable=False` but create an alternate pathway from which the LMS streams the file to the learner? Or keep `downloadable=True` but somehow apply a permissions gate before the learner gets the download URL ?

Reply:
> Would `learner_downloadable` be more a accurate name?

Yeah, probably. I guess it's more specifically that it shouldn't be downloadable in the learning experience, regardless of your permissions. Everything is downloadable in the authoring experience. But `learner_downloadable` is closer to the right nuance than what I have.

> Also, I don't really understand; with this approach, if there is some PDF asset that the LMS wants to make available only to enrolled learners, how would we do that? Set `downloadable=False` but create an alternate pathway from which the LMS streams the file to the learner? Or keep `downloadable=True` but somehow apply a permissions gate before the learner gets the download URL ?

In this scenario, the `learner_downloadable` flag would be True, because PDFs are intended for download. But that's not sufficient–something on the LMS side decides whether this is freely downloadable by everyone or only by people enrolled in the course. The `learner_downloadable` flag is only False for things that students should never get (it's not even a permissions option–that type of resource should just never be downloadable via the LMS), like `python_lib.zip`, the OLX of various XBlocks, etc.

Reply:
So the goal here is to indicate whether the content is to be used by the system on behalf of the instructor instead of being made directly available to the learner?  I think explicitly putting the example of `python_lib.zip` in the comment will help people grok this more quickly.

Also, is that the only thing that would be marked as `False` for now or are there other things that currently fall into this category?

Reply:
> So the goal here is to indicate whether the content is to be used by the system on behalf of the instructor instead of being made directly available to the learner? I think explicitly putting the example of `python_lib.zip` in the comment will help people grok this more quickly.

Okay, I'll add that example to the comment in my next update.

> Also, is that the only thing that would be marked as `False` for now or are there other things that currently fall into this category?

All XBlock source OLX falls into this category, because it often has answers baked into it. So learners should never be able to download a ProblemBlock's OLX directly. I put this example in the code comments because I figured it would be the most common use case people would run into. I could imagine other use cases that are more niche–for instance, some MIT courses have LaTeX versions of their problems that they compile into OLX, and it would make sense to store those next to the problem OLX but prevent students from downloading them.